signature/pipeline_invariants_test.go

```diff
@@ -0,0 +1,179 @@
+			Plugins: nil, // nil plugins - this should become nil
+			Secrets: nil, // nil secrets - this should become nil
+		},
+		RepositoryURL: "https://github.com/example/repo",
+		OuterEnv:      map[string]string{"PIPELINE_VAR": "value"},
+	}
+
+	fields, err := step.SignedFields()
+	if err != nil {
+		t.Fatalf("SignedFields() error = %v", err)
+	}
+
+	// Check that secrets field is NOT present when empty (for backward compatibility)
+	if _, has := fields["secrets"]; has {
+		t.Errorf("SignedFields() should not include 'secrets' field when empty for backward compatibility")
+	}
+}
+
+func TestCommandStepWithInvariants_ValuesForFields_WithSecrets(t *testing.T) {
+	t.Parallel()
+
+	dbUrl := "DATABASE_URL"
+	step := commandStepWithInvariants{
+		CommandStep: pipeline.CommandStep{
+			Command: "echo hello",
+			Env:     map[string]string{"FOO": "bar"},
+			Plugins: pipeline.Plugins{},
+			Secrets: pipeline.Secrets{
+				{Key: "DATABASE_URL", EnvironmentVariable: &dbUrl},
+			},
+		},
+		RepositoryURL: "https://github.com/example/repo",
+		OuterEnv:      map[string]string{"PIPELINE_VAR": "value"},
+	}
+
+	fields := []string{"command", "env", "plugins", "matrix", "repository_url", "secrets"}
+	values, err := step.ValuesForFields(fields)
+	if err != nil {
+		t.Fatalf("ValuesForFields() error = %v", err)
+	}
+
+	// Check that secrets value is included
+	if _, has := values["secrets"]; !has {
+		t.Errorf("ValuesForFields() missing 'secrets' field")
+	}
+
+	// Check that secrets field contains expected data
+	secrets, ok := values["secrets"].(pipeline.Secrets)
+	if !ok {
+		t.Errorf("ValuesForFields() 'secrets' field is not pipeline.Secrets type, got %T", values["secrets"])
+	}
+
+	if len(secrets) != 1 {
+		t.Errorf("ValuesForFields() expected 1 secret, got %d", len(secrets))
+	}
+
+	if secrets[0].Key != "DATABASE_URL" {
+		t.Errorf("ValuesForFields() expected secret key 'DATABASE_URL', got %q", secrets[0].Key)
+	}
+}
+
+func TestCommandStepWithInvariants_ValuesForFields_MissingSecretsField(t *testing.T) {
+	t.Parallel()
+
+	dbUrl := "DATABASE_URL"
+	step := commandStepWithInvariants{
+		CommandStep: pipeline.CommandStep{
+			Command: "echo hello",
+			Env:     map[string]string{"FOO": "bar"},
+			Plugins: pipeline.Plugins{},
+			Secrets: pipeline.Secrets{
+				{Key: "DATABASE_URL", EnvironmentVariable: &dbUrl},
+			},
+		},
+		RepositoryURL: "https://github.com/example/repo",
+		OuterEnv:      map[string]string{"PIPELINE_VAR": "value"},
+	}
+
+	// Request fields without secrets - should fail
+	fields := []string{"command", "env", "plugins", "matrix", "repository_url"}
+	_, err := step.ValuesForFields(fields)
+	if err == nil {
+		t.Fatalf("ValuesForFields() expected error when secrets field not requested but step has secrets")
+	}
+
+	expectedError := "one or more required fields are not present: [secrets]"
+	if err.Error() != expectedError {
+		t.Errorf("ValuesForFields() expected error %q, got %q", expectedError, err.Error())
+	}
```

Comment:
Good use of `%q` 👍

Checking the error string is acceptable since we control the implementation, but is prone to being a "change detector". At some point maybe we should introduce a new error type that can be [checked more semantically](https://google.github.io/styleguide/go/decisions#test-error-semantics) than  we can via a string comparison. 

Other comment here is ["got before want"](https://google.github.io/styleguide/go/decisions#got-before-want):

```suggestion
	wantErr := "one or more required fields are not present: [secrets]"
	if err.Error() != wantErr {
		t.Errorf("ValuesForFields() error = %q, want %q", err.Error(), wantErr)
	}
```